### PR TITLE
Updated Big Tent to support generic data records

### PR DESCRIPTION
**Pull request by dpp**
### 💻 Description of Change(s) (w/ context)
Changed from Big Tent only storing `ItemMetaData` as records to supporting generic types
so that Big Tent can be a more generic graph database

### 🧠 Rationale Behind Change(s)
Moving Big Tent from a "[Spice SaLAD](https://spicesalad.org)" only database to being a more
generic graph database.

### 📝 Test Plan
Tested the code against a series of "clusters" generated by [Goat Rodeo](https://goatrodeo.cc) and
the system functioned normally

### 📜 Documentation
Documented some of the types.

### 💣 Quality Control
(All items must be checked before a PR is merged)
Did you…
- [ No] Mention an issue number in the PR title?
- [ x] Update the version # in the build file?
- [ x] Create new and/or update relevant existing tests?
- [ x] Create or update relevant documentation and/or diagrams?
- [ x] Comment your code?
- [ x] Fix any stray verbose logging (removing, or moving to debug / trace level)?

### Before Merging…
- Make sure the Quality Control boxes are all ticked
- Make sure any open comments or conversations on the PR are resolved

### CODE REVIEW

File: src/cluster_writer.rs [hunk 1]
```diff
@@ -93,7 +94,10 @@ impl ClusterWriter {
     self.previous_position
   }
 
-  pub fn write_item(&mut self, mut item: Item) -> Result<()> {
+  pub fn write_item<MDT>(&mut self, mut item: Item<MDT>) -> Result<()>
+  where
+    for<'de2> MDT: Deserialize<'de2> + Serialize + PartialEq + Clone + Mergeable + Sized,
```
Comment (earldouglas) [inline src/cluster_writer.rs:99]: Can we create a single trait to encapsulate all of these?

**Comment by earldouglas** [comment]
Can we write some tests for this?

File: src/data_file.rs [hunk 2]
```diff
@@ -24,22 +24,31 @@ pub struct DataFileEnvelope {
...
-pub struct DataFile {
+pub struct DataFile<MDT>
+where
+  for<'de2> MDT:
+    Deserialize<'de2> + Serialize + PartialEq + Clone + Mergeable + Sized + Send + Sync + 'static,
+{
   pub envelope: DataFileEnvelope,
   pub file: Arc<Mutex<BufReader<File>>>,
   pub data_offset: u64,
+  _phantom: Option<MDT>,
```
Comment (earldouglas) [inline src/data_file.rs:35]: Is this a convention in Rust?  If not, can we comment about Rust and phantom types?

File: src/rodeo.rs [hunk 3]
```diff
@@ -770,8 +790,8 @@ fn test_files_in_dir() {
     cluster.get_index().unwrap(); // should be from cache
     let time2 = Instant::now().duration_since(start);
     assert!(
-      time > Duration::from_millis(60),
-      "Building the initial index should take more than 60ms, but took {:?}",
+      time > Duration::from_millis(20),
```
Comment (earldouglas) [inline src/rodeo.rs:793]: This approach works well enough, but let's comment on how we picked 20ms.

File: src/structs.rs [hunk 4]
```diff
@@ -4,165 +4,182 @@ use serde::{Deserialize, Serialize};
...
+  pub fn is_alias(&self) -> bool {
+    self.connections.iter().any(|v| v.0 == EdgeType::AliasTo)
+  }
+  pub fn remove_references(&mut self) {
+    self.reference = Item::<MDT>::NOOP;
+    self.merged_from = BTreeSet::new();
+  }
+
+  // tests two items... they are the same if they have the same
+  // `identifier`, `connections`, `metadata`, `file_size`, `version`,
```
Comment (earldouglas) [inline src/structs.rs:119]: s/file_size//

**Review by earldouglas — DISMISSED** [review]

File: src/cluster_writer.rs (hunk 1 above)
Comment (dpp) [reply src/cluster_writer.rs:99]: Addressed

File: src/data_file.rs (hunk 2 above)
Comment (dpp) [reply src/data_file.rs:35]: Added comment and the correct Rust pattern: `PhantomData`

File: src/main.rs [hunk 5]
```diff
@@ -123,7 +127,7 @@ fn main() -> Result<()> {
   match (&args.rodeo, &args.conf, &args.fresh_merge) {
     (Some(rodeo), None, v) if v.len() == 0 => run_rodeo(rodeo, &args)?,
     (None, Some(_conf), v) if v.len() == 0 => run_full_server(args)?,
-    (None, None, v) if v.len() > 0 => run_merge(v.clone(), args)?,
+    (None, None, v) if v.len() > 0 => run_merge::<ItemMetaData>(v.clone(), args)?,
```
[the remark this answers is not in the document]
Comment (dpp) [reply src/main.rs:130]: :+1:

File: src/rodeo.rs (hunk 3 above)
Comment (dpp) [reply src/rodeo.rs:793]: Commented

**Review by earldouglas — APPROVED** [review]

File: Cargo.toml [hunk 6]
```diff
@@ -2,9 +2,23 @@
 name = "bigtent"
 version = "0.5.0"
 edition = "2021"
+authors = ["David Pollak <feeder.of.the.bears@gmail.com>"]
```
Comment (bwmcadams) [inline Cargo.toml:5]: should this be a spice labs address? ext-engineering?

**Review by bwmcadams — CHANGES_REQUESTED** [review]

**Pull request merged**